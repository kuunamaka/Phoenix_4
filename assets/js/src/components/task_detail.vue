<template>
  <table v-if="task">
    <h2>Task Details</h2>
    <div>
      <b>Task:</b> {{ task.name }} (by: <b>{{ task.assignee }}</b
      >)
    </div>
    <div>
      <b>Is_finished?:</b> {{ task.is_finished }}, <b>finished_at:</b>
      {{ task.finished_at }},
      <button
        v-if="task.is_liked"
        @click="deleteLike(task)"
        class="fas fa-heart"
      ></button>
      <button v-else @click="createLike(task)" class="far fa-heart"></button>
      <b>{{ task.favorited_users.length }}</b>
    </div>
    <div v-if="errors.length">
      <p v-for="error in errors" :key="error" class="error">{{ error }}</p>
    </div>
    <div>
      Comment: <b>{{ task.comment }}</b>
    </div>
    <div>
      <input v-model="comment" type="text" />
      <button @click="createComment(comment)" class = "comment">Add Comment</button>
    </div>
    <a href="/tasks" class="button">Exit</a>
  </table>
</template>

<script>
import axios from "axios";
export default {
  name: "task_detail",
  data() {
    return {
      task: null,
      comment: null,
      errors: [],
    };
  },
  methods: {
    async createLike(task) {
      await axios.post(`/api/tasks/${this.task.id}/users`, {
        task: {
          name: this.name,
          id: this.task.id,
        },
      });
      window.location.href = `/tasks/${this.task.id}`;
    },
    async deleteLike(task) {
      await axios.delete(`/api/tasks/${this.task.id}/users/1`, {
        task: {
          name: this.name,
          id: this.task.id,
        },
      });
      window.location.href = `/tasks/${this.task.id}`;
    },
    async createComment(comment) {
      this.errors = [];
      if (!this.comment) {
        this.errors.push("Comments cannot be blank");
      }
      if (this.errors.length !== 0) {
        return;
      }
      await axios.post(`/api/tasks/${this.task.id}/comments`, {
        comment: {
          body: this.comment,
          task_id: this.task.id,
        }
      });
      window.location.href = `/tasks/${this.task.id}`;
    },
  },
  async mounted() {
    const task_id = window.location.pathname.split("/")[2];
    const resp = await axios.get(`/api/tasks/${task_id}`);
    this.task = resp.data;
  },
};
</script>
