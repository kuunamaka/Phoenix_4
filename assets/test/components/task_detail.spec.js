import { shallowMount } from "@vue/test-utils";
import task_detail from "@/src/components/task_detail.vue";

const axios = require("axios");
const MockAdapter = require("axios-mock-adapter");
const mock = new MockAdapter(axios);

mock.onGet("/api/tasks/1").reply(200, {
  id: 1,
  name: "Test",
  is_finished: false,
  assignee_id: 1,
});

window = Object.create(window);
const pathname = "/tasks/1";

Object.defineProperty(window, "location", {
  value: {
    pathname,
    href: `http://localhost:4001${pathname}`,
  },
});

describe("task_detail.vue", () => {
  const wrapper = shallowMount(task_detail);
  const vm = wrapper.vm;

  it("renders entire component correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe("createLike/1", () => {
    it("redirect to /tasks/${task.id}", async () => {
      mock
        .onPost("/api/tasks/1/users", { task: { id: 1, name: "Test" } })
        .reply(200);
      await wrapper.setData({ id: 1, name: "Test" });

      await vm.createLike({ id: 1, name: "Test" });

      expect(window.location.href).toEqual("/tasks/1");
    });
  });

  describe("deleteLike/1", () => {
    it("redirect to /tasks/${task.id}", async () => {
      mock
        .onDelete("/api/tasks/1/users/1", { task: { id: 1, name: "Test" } })
        .reply(200);
      await wrapper.setData({ id: 1, name: "Test" });

      await vm.deleteLike({ id: 1, name: "Test" });

      expect(window.location.href).toEqual("/tasks/1");
    });
  });

  describe("createComment", () => {
    it("redirect to /tasks/${task.id}", async () => {
      mock
        .onPost("/api/tasks/1/comments", {
          comment: {
            body: "Test Comment",
            task_id: 1
          }
        })
        .reply(200);
      await wrapper.setData({ body: "Test Comment", task_id: 1 });
      
      await vm.createComment({ body: "Test Comment", task_id: 1 });

      expect(window.location.href).toEqual("/tasks/1");
    });

    it("show validate error messages", async () => {
      mock
        .onPost("/api/tasks/1/comments", {
          comment: {
            body: "Test Comment",
            task_id: 1,
          },
        })
        .reply(404);
      await wrapper.setData({ body: "", task_id: 1 });
  
      await vm.createComment();
  
      expect(wrapper.find("p").element).toMatchSnapshot();
    });
  });
});
