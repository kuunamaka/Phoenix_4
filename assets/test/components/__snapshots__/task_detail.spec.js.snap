// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`task_detail.vue createComment show validate error messages 1`] = `undefined`;

exports[`task_detail.vue renders entire component correctly 1`] = `<!---->`;
